controller/kubernetes-pdb-controller.go

```diff
@@ -0,0 +1,222 @@
+	*baseController
+
+	// which namespace controller is running with
+	namespace string
+	// use as the OwnerID of the controller
+	controllerID string
+
+	kubeClient clientset.Interface
+
+	ds *datastore.DataStore
+
+	cacheSyncs []cache.InformerSynced
+}
+
+func NewPDBController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string) *PDBController {
+
+	pc := &PDBController{
+		baseController: newBaseController("kubernetes-pdb", logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+
+		ds:         ds,
+		kubeClient: kubeClient,
+	}
+
+	ds.DeploymentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    pc.enqueueDeployment,
+		UpdateFunc: func(old, cur interface{}) { pc.enqueueDeployment(cur) },
+		DeleteFunc: pc.enqueueDeployment,
+	})
+	pc.cacheSyncs = append(pc.cacheSyncs, ds.DeploymentInformer.HasSynced)
+
+	ds.VolumeInformer.AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    pc.enqueueVolume,
+		UpdateFunc: func(old, cur interface{}) { pc.enqueueVolume(cur) },
+		DeleteFunc: pc.enqueueVolume,
+	}, 0)
+	pc.cacheSyncs = append(pc.cacheSyncs, ds.VolumeInformer.HasSynced)
+
+	return pc
+}
+
+func (pc *PDBController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer pc.queue.ShutDown()
+
+	pc.logger.Infof("Starting Kubernetes PDB controller")
+	defer pc.logger.Infof("Shut down Kubernetes PDB controller")
+
+	if !cache.WaitForNamedCacheSync(pc.name, stopCh, pc.cacheSyncs...) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(pc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (pc *PDBController) worker() {
+	for pc.processNextWorkItem() {
+	}
+}
+
+func (pc *PDBController) processNextWorkItem() bool {
+	key, quit := pc.queue.Get()
+	if quit {
+		return false
+	}
+	defer pc.queue.Done(key)
+	err := pc.syncHandler(key.(string))
+	pc.handleErr(err, key)
+	return true
+}
+
+func (pc *PDBController) handleErr(err error, key interface{}) {
+	if err == nil {
+		pc.queue.Forget(key)
+		return
+	}
+
+	pc.logger.WithError(err).Warnf("Error syncing PDB for %v", key)
+	pc.queue.AddRateLimited(key)
+}
+
+func (pc *PDBController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: failed to sync PDB for %v", pc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != pc.namespace {
+		return nil
+	}
+
+	if _, ok := targetingDeployment[name]; !ok {
+		return nil
+	}
+	return pc.reconcile(name)
+}
+
+func (pc *PDBController) reconcile(name string) (err error) {
+	pdbName := name
+	pdb, err := pc.ds.GetPDBRO(pdbName)
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	deployment, err := pc.ds.GetDeployment(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return deletePDBObject(pc.logger, pc.ds, pdb)
+		}
+		return err
+	}
```

Comment:
When will we have this case?

Reply:
Yeah, I think during the uninstallation I guess

Reply:
I see, but from the code, it seems there will be no chance to make this happen since the manager daemonset should be deleted first before the other deployments get deleted? However, this could avoid unexpected deletion somehow to ensure PDB depending on these components. 

[/controller/uninstall_controller.go#L281-L308](https://github.com/longhorn/longhorn-manager/blob/65e6555ce61e279ce1e25dbbcecff5b1670f168e/controller/uninstall_controller.go#L281-L308)
```go
	// A race condition exists where manager may attempt to recreate certain
	// CRDs after their deletion. We must delete manager and do a final check
	// for CRDs, just in case.
	if waitForUpdate, err := c.deleteManager(); err != nil || waitForUpdate {
		return err
	}

	if waitForUpdate, err := c.deleteDriver(); err != nil || waitForUpdate {
		return err
	}

	// We set gracePeriod=0s because there is no possibility of graceful
	// cleanup without a running manager.
	gracePeriod = 0 * time.Second
	if waitForUpdate, err := c.deleteCRDs(); err != nil || waitForUpdate {
		return err
	}

	deployments := []string{
		types.LonghornRecoveryBackendDeploymentName,
		types.LonghornAdmissionWebhookDeploymentName,
		types.LonghornConversionWebhookDeploymentName,
	}
	for _, deployment := range deployments {
		if waitForUpdate, err := c.deleteDeployment(deployment); err != nil || waitForUpdate {
			return err
		}
	}

```